sos/collector/__init__.py

```diff
@@ -0,0 +1,1091 @@
+            self.exit('Could not connect to master node. Aborting...', 1)
+
+    def determine_cluster(self):
+        """This sets the cluster type and loads that cluster's cluster.
+
+        If no cluster type is matched and no list of nodes is provided by
+        the user, then we abort.
+
+        If a list of nodes is given, this is not run, however the cluster
+        can still be run if the user sets a --cluster-type manually
+        """
+        checks = list(self.clusters.values())
+        for cluster in self.clusters.values():
+            checks.remove(cluster)
+            cluster.master = self.master
+            if cluster.check_enabled():
+                cname = cluster.__class__.__name__
+                self.log_debug("Installation matches %s, checking for layered "
+                               "profiles" % cname)
+                for remaining in checks:
+                    if issubclass(remaining.__class__, cluster.__class__):
+                        rname = remaining.__class__.__name__
+                        self.log_debug("Layered profile %s found. "
+                                       "Checking installation"
+                                       % rname)
+                        remaining.master = self.master
+                        if remaining.check_enabled():
+                            self.log_debug("Installation matches both layered "
+                                           "profile %s and base profile %s, "
+                                           "setting cluster type to layered "
+                                           "profile" % (rname, cname))
+                            cluster = remaining
+                            break
+                self.cluster = cluster
+                self.cluster_type = cluster.name()
+                self.commons['cluster'] = self.cluster
+                self.ui_log.info(
+                    'Cluster type set to %s' % self.cluster_type)
+                break
+
+    def get_nodes_from_cluster(self):
+        """Collects the list of nodes from the determined cluster cluster"""
+        if self.cluster_type:
+            nodes = self.cluster._get_nodes()
+            self.log_debug('Node list: %s' % nodes)
+            return nodes
+        return []
+
+    def reduce_node_list(self):
+        """Reduce duplicate entries of the localhost and/or master node
+        if applicable"""
+        if (self.hostname in self.node_list and self.opts.no_local):
+            self.node_list.remove(self.hostname)
+        for i in self.ip_addrs:
+            if i in self.node_list:
+                self.node_list.remove(i)
+        # remove the master node from the list, since we already have
+        # an open session to it.
+        if self.master:
+            for n in self.node_list:
+                if n == self.master.hostname or n == self.opts.master:
+                    self.node_list.remove(n)
```

Comment:
When running:

    python3 ./bin/sos collect --nodes=$(hostname -f),some.existing.system --batch

then this piece of code removes `$(hostname -f)` from the `self.node_list` (confirmed by adding some debugs). Why is it so? From a user experience, it is unexpected.

Reply:
OK, the reason why `sos report` was skipped from the local host was, that `~/.ssh/known_hosts` contained a valid record for `$(hostname)` but not `localhost`.

Tricky small issue, not blocker for the merge but worth fixing later on, imho.